Open-source kafka library

jwoudenberg-nri opened the pull request:
- [x] meta-fields filled in `package.yaml` (do not rely on package.yaml support)
- [x] Add license and changelog
- [x] version bounds for dependencies in `package.yaml`
- [x] start kafka in tests so CI can pass
- [x] no internal functions exposed
- [x] reviewed exposed constructors (maybe avoid if it would likely result in backwards-incompatible breakage in the future)
- [x] reviewed number of modules
- [x] code compiles and tests run, CI passes
- [x] `cabal haddock` reports 100% doc coverage, doc comments use haddock syntax
- [x] use open-sourced version of code in private repo succesfully

jwoudenberg committed:
Open-source kafka library

Co-authored-by: Christoph Hermann <schtoeffel@gmail.com>
Co-authored-by: Michael Glass <me@mike.is>

jwoudenberg committed:
Add metadata fields to kafka package.yaml

michaelglass committed:
add build dependencies

michaelglass committed:
add kafka entry to makefile

michaelglass committed:
rename to nri-kafka

michaelglass committed:
install hw-kafka-client in nix

michaelglass committed:
remove depedency on CLI

michaelglass committed:
start working on getting tests working in CI

michaelglass committed:
start kafka in tests

michaelglass committed:
document Kafka module

michaelglass committed:
document Kafka.Settings

michaelglass committed:
document Kafka.Worker

michaelglass committed:
document Worker.Settings

michaelglass committed:
reduce number of exported modules, and add missing documentation

michaelglass committed:
don't run tests until kafka starts

michaelglass committed:
stop kafka before we start it (just in case)

michaelglass committed:
start zookeeper

michaelglass committed:
cabal files all point explicitly to our readme

jwoudenberg committed:
Separate internal and external test helpers

The test helpers we use for testing the library itself aren't amazing.
They flake, are pretty slow, and mess up test runner output. We
recommend users of this library use `stub` to make it so tests don't
need to run a real Kafka, or test against a real environment.

jwoudenberg committed:
Remove redis-content from kafka CHANGELOG

jwoudenberg committed:
Don't expose processWithoutShutdownEnsurance

This is an internal helper function.

jwoudenberg committed:
Replace Description with TopicSubscriptions type

TopicSubscriptions is a more descriptive name, but we have to loose the
`groupId` field to make it not be a lie.

jwoudenberg committed:
Rename ProcessResult to SeekCmd

This name makes more sense with the change that's coming in the next
commit.

jwoudenberg committed:
Add subscription creating helpers

This serves a couple of purposes:
- Don't expose some constructors so we can make changes in the future
  without major version bumps.
- No need for library users to use a lot of wrapper types, without
  loosing benefit of those types inside the library code.

jwoudenberg committed:
Drop wrapper type exports

Now that we have the subscription construction functions we don't use
these types anymore anywhere on the API (as far as I'm aware).

jwoudenberg committed:
Add KAFKA_ prefix to all settings vars

michaelglass commented on nri-kafka/src/Kafka/Worker/Internal.hs:
```diff
@@ -46,11 +46,10 @@ data State = State
-    topic :: Kafka.Topic,
+-- | The topics this worker should subscribe too. At the moment this library
+-- only supports subscribing too a single topic.
+data TopicSubscription = TopicSubscription
```
better name, too!

michaelglass commented on nri-kafka/src/Kafka/Worker/Settings.hs:
```diff
@@ -148,3 +154,13 @@ decoderOnProcessMessageFailure =
+    Environment.Variable
+      { Environment.name = "KAFKA_GROUP_ID",
+        Environment.description = "Identifies the consumer group this worker is part off",
+        Environment.defaultValue = "procrastinators"
```
maybe we shouldn't have a default value at all?

michaelglass committed:
move group id from settings to explicit argument
we expect settings to have a default value. GroupID should never have a default value because it would make it easy to deploy two sets of different workers with the same groupID. That would lead kafka brokers to distribute partitions wrong, as well as give the wrong offsetted-messages to each worker

michaelglass commented on nri-kafka/src/Kafka/Worker.hs:
```diff
@@ -22,7 +22,7 @@ module Kafka.Worker
     Consumer.TopicName (..),
     Consumer.ConsumerRecord (..),
-    Partition.ProcessResult (..),
+    Partition.SeekCmd (..),
```
much better

michaelglass commented on nri-kafka/src/Kafka/Worker.hs:
```diff
@@ -15,18 +15,26 @@
-    Internal.TopicSubscription (..),
-    Partition.MessageCallback (Partition.MessageCallback),
+
+    -- * Settings
```
✨

michaelglass commented on nri-kafka/src/Kafka/Worker/Internal.hs:
```diff
@@ -54,6 +55,119 @@ data TopicSubscription = TopicSubscription
+subscription ::
+  (Aeson.FromJSON msg, Aeson.ToJSON msg) =>
+  Text ->
+  (Envelope msg -> Task Text ()) ->
```
✨

michaelglass commented on nri-kafka/src/Kafka/Worker/Internal.hs:
```diff
@@ -203,7 +316,7 @@ rebalanceCallback skipOrNot observability callback offsetSource state consumer r
             fetchResult <- Task.attempt log (fetch newPartitions)
             case fetchResult of
-              Err err -> Exception.throwIO err
+              Err err -> Exception.throwString (Text.toList err)
```
why string?

michaelglass commented on nri-kafka/src/Kafka/Worker.hs:
```diff
@@ -26,19 +26,9 @@ module Kafka.Worker
     Internal.Envelope (topicName, partitionId, messageOffset, payload),
     Partition.SeekCmd (..),
-
-    -- * Wrapper types
```
🤘

michaelglass commented on nri-kafka/src/Kafka/Worker/Settings.hs:
```diff
@@ -142,7 +142,7 @@ decoderOnProcessMessageFailure :: Environment.Decoder SkipOrNot
   Environment.variable
     Environment.Variable
-      { Environment.name = "SKIP_ON_PROCESS_MESSAGE_FAILURE",
+      { Environment.name = "KAFKA_SKIP_ON_PROCESS_MESSAGE_FAILURE",
```
❤️

michaelglass commented on nri-kafka/src/Kafka/Worker.hs:
```diff
@@ -26,19 +26,9 @@ module Kafka.Worker
-
-    -- * Wrapper types
-    Internal.OffsetSource (..),
-    Consumer.ConsumerGroupId (..),
```
I believe we still need to export this now that it's an argument

michaelglass committed:
update readme

michaelglass committed:
Kafka.Worker.process takes a string instead of a groupId

michaelglass committed:
subscription doesn't need the envelope

michaelglass committed:
rename Envelope to OffsetParams

michaelglass committed:
rename initialOffsets to fetchOffsets and pass in OffsetParams

michaelglass committed:
avoid ambiguous Dict.Int Int

jwoudenberg committed:
Message ParamOffsets type

The core idea here is that we don't really need the topic name to be
part of the ParamOffsets type. The ParamOffsets type is used in the
context of setting up a subscription for a single topic, so the caller
already knows which topic we're talking about.

Dropping the topic name makes it so our fetch offsets function only
needs to provide the partition ids we're interested in, for which we can
use plain Ints. This allows us to specialize the `ParamOffsets` type for
representing a particular partition offset, and include an offset field.

jwoudenberg committed:
Make message key and payload optional

Kafka itself doesn't require them, so let's not either.

jwoudenberg committed:
Add helper functions for constructing a message

This provides some nice opportunities for documentation, and removes the
need to use wrapper types in the API to distinguish arguments (like the
topic name and key).

jwoudenberg committed:
Don't expose Kafka.Topic and Kafka.Key types

Our public API's are no longer using these types, so library users have
no use for them.

jwoudenberg committed:
Remove OnDelivery type alias

It was only used in a single place, and we weren't exposing it.

jwoudenberg committed:
Improve usability of payload function

jwoudenberg committed:
Don't expose error type

We're not using these constructors at the moment in our client code.
Maybe a use for them will appear in the future, but we don't know yet
what that will look like and how we'd solve it, when it happens.

jwoudenberg committed:
Expose PartitionOffset constructor

Clients need to be able to construct values of this type.

The risk of this type gaining additional constructors or fields (which
would result in backwards-compatiblity breakig changes) is low, so
exposing constructor here seems fine.

ghost commented:
I got the current version compiling against our private repo.

I'm going to merge this so we can move forward using this in our private repo, but not publish to hackage yet in case there's additional changes we'd like to make.

michaelglass commented on nri-kafka/src/Kafka/Worker/Internal.hs:
```diff
@@ -108,10 +108,10 @@ subscription topic callback =
   (Aeson.FromJSON msg, Aeson.ToJSON msg) =>
   Text ->
-  Task Text (Dict.Dict Int Int) ->
+  ([OffsetParams] -> Task Text (Dict.Dict Int Int)) ->
```
🚨over-engineering🚨 perhaps the destination database will determine where to pull / put the data based on partition id?

agree we should drop it until it becomes an issue.

michaelglass commented on nri-kafka/src/Kafka/Worker/Internal.hs:
```diff
@@ -102,14 +102,17 @@ subscription topic callback =
   Text ->
-  ([OffsetParams] -> Task Text (Dict.Dict OffsetParams Int)) ->
-  (OffsetParams -> Int -> msg -> Task Text Partition.SeekCmd) ->
+  ([Int] -> Task Text (List PartitionOffset)) ->
```
👍

michaelglass commented on nri-kafka/src/Kafka.hs:
```diff
@@ -144,8 +144,10 @@ topic :: Internal.Msg -> Text
 -- | The payload of a message. This function might sometimes be useful in tests.
-payload :: Internal.Msg -> Maybe Aeson.Value
-payload msg = Maybe.map Aeson.toJSON (Internal.payload msg)
+payload :: (Aeson.FromJSON a) => Internal.Msg -> Maybe a
```
👍

michaelglass commented:
thank you!
